mono/tests/Makefile.am

```diff
@@ -1078,13 +1133,35 @@ TESTSAOT_IL=$(TESTS_IL:.exe=.exe$(PLATFORM_AOT_SUFFIX))
 TESTSAOT_BENCH=$(TESTS_BENCH:.exe=.exe$(PLATFORM_AOT_SUFFIX))
 TESTSAOT_STRESS=$(TESTS_STRESS:.exe=.exe$(PLATFORM_AOT_SUFFIX))
 TESTSAOT_GSHARED=$(TESTS_GSHARED:.exe=.exe$(PLATFORM_AOT_SUFFIX))
+#FIXME
+#TESTSAOT_TAILCALL=$(TEST_TAILCALL:.exe=.exe$(PLATFORM_AOT_SUFFIX))
 endif
 
 EXTRA_DIST=test-driver test-runner.cs $(TESTS_CS_SRC) $(TESTS_IL_SRC) \
-	$(TESTS_BENCH_SRC) $(TESTS_STRESS_SRC) stress-runner.pl
+	$(TESTS_BENCH_SRC) $(TESTS_STRESS_SRC) stress-runner.pl \
+	$(TEST_TAILCALL_CS_SRC) \
+	$(TEST_TAILCALL_IL_SRC)
+
+# This test has extra sources.
+tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyWithPrefix.exe: \
+		tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyTransparentLibraryWithPrefix.il \
+		tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyVerifiableLibraryWithPrefix.il
 
 %.exe: %.il
-	$(ILASM) -out:$@ $<
+	$(ILASM) -out:$@ $^
+
+# Define I to empty to enable incremental.
+I ?= -not-incremental
+
+# FIXME missing dependencies on BCL (just Console.WriteLine)
+%.exe_test_tailcall_done: %.exe Makefile $(srcdir)/../../runtime/mono-wrapper ../mini/mono
```

Comment:
Let's just have this make target:

```
test-tailcall: $(TESTS_TAILCALL)
    MONO_DEBUG=test-tailcall-require $(TOOLS_RUNTIME) --debug $(TEST_RUNNER) -j a --testsuite-name "runtime-tailcall" --timeout 300 --disabled "$(DISABLED_TESTS)" $<
```

Reply:
There are several problems with your proposal that I carefully avoided.
  - You take away optional incrementality.
  - You add nested parallelism. Nested parallelism is too much parallelism. Only make should control this.
  - By removing the -v and grep, you take away verification that any tail calls occur. Hello world would pass, with no tail calls.
 - By running the tests instead of using --compile-all, you induce a variety of problems. The tests fail for a variety of reasons. They can be windows specific for example (actual case). --compile-all gets us, mostly, JIT coverage, w/o worrying if the code runs. As well, if there is a failure, the failure mode is understood and occurs right away (I've seen it). The only problem with --compile-all is that it skips generics, so we are losing some coverage perhaps. But meanwhile I'm writing more deliberate tests.
  - -compile-all runs in a fairly deterministic finite time, w/o depending on a timeout. That is good.
  - We could go threw the tests and determine which runs are runnable, and only -compile-all or skip entirely the ones that aren't runnable, but blind blanket -compile-all is not a terrible strategy.

On the flip side, throwing test-tailcall-require at random tests isn't a bad idea. There are actually other tailcalls in the test suite already in non-obvious places (the EvenOdd test).

Reply:
I effectively missed the `--compile-all` but it's pretty easy to fix:

```
test-tailcall: $(TESTS_TAILCALL)
    MONO_DEBUG=test-tailcall-require $(TOOLS_RUNTIME) --debug $(TEST_RUNNER) -j a  --runtime-args "--compileall" --testsuite-name "runtime-tailcall" --timeout 300 --disabled "$(DISABLED_TESTS)" $<
```

Reply:
That not verify there are any tail calls. Just assume there are?

Reply:
We pass `MONO_DEBUG=test-tailcall-require` which guarantees the `tail.` prefix are respected. If you are talking about assuming there are `tail.` prefix in the test cases: they are out test cases, we control their content.

Reply:
Right the second part. Not everything in the initial set had "tail." in it. The grep failed those tests appropriately.

Reply:
Then we should add the `tail.` appropriately. The advantage of using the `test-runner` is that you get integration with our CI for free. Just failing the check makes it hard - even for us - to figure out what exactly went wrong, so by using the `test-runner` and having the proper output integrated with CI, you make it easy to know what test failed and why.